tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -28,62 +34,214 @@
+
+        private final Deque<OpenSpan> stack;
+
+        private Sampled(ArrayDeque<OpenSpan> stack, String traceId) {
+            super(traceId);
+            this.stack = stack;
+        }
+
+        private Sampled(String traceId) {
+            this(new ArrayDeque<>(), traceId);
+        }
+
+        @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored") // Sampled traces cannot optimize this path
+        void fastStartSpan(String operation, String parentSpanId, SpanType type) {
+            startSpan(operation, parentSpanId, type);
+        }
+
+        @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored") // Sampled traces cannot optimize this path
+        void fastStartSpan(String operation, SpanType type) {
+            startSpan(operation, type);
+        }
+
+        @Override
+        void push(OpenSpan span) {
+            stack.push(span);
+        }
+
+        @Override
+        Optional<OpenSpan> top() {
+            return stack.isEmpty() ? Optional.empty() : Optional.of(stack.peekFirst());
+        }
+
+        @Override
+        Optional<OpenSpan> pop() {
+            return stack.isEmpty() ? Optional.empty() : Optional.of(stack.pop());
+        }
+
+        @Override
+        boolean isEmpty() {
+            return stack.isEmpty();
+        }
+
+        @Override
+        boolean isObservable() {
+            return true;
+        }
+
+        @Override
+        Trace deepCopy() {
+            return new Sampled(new ArrayDeque<>(stack), getTraceId());
+        }
+
+        @Override
+        public String toString() {
+            return "Trace{stack=" + stack + ", isObservable=true, traceId='" + getTraceId() + "'}";
+        }
     }
 
-    @Override
-    public String toString() {
-        return "Trace{stack=" + stack + ", isObservable=" + isObservable + ", traceId='" + traceId + "'}";
+    private static final class Unsampled extends Trace {
+        private int depth;
+
+        private Unsampled(int depth, String traceId) {
+            super(traceId);
+            this.depth = depth;
+            validateDepth();
+        }
+
+        private Unsampled(String traceId) {
+            this(0, traceId);
+        }
+
+        @Override
+        void fastStartSpan(String operation, String parentSpanId, SpanType type) {
+            fastStartSpan(operation, type);
+        }
+
+        @Override
+        void fastStartSpan(String operation, SpanType type) {
+            depth++;
+        }
+
+        @Override
+        void push(OpenSpan span) {
+            depth++;
+        }
+
+        @Override
+        Optional<OpenSpan> top() {
+            return Optional.empty();
+        }
+
+        @Override
+        Optional<OpenSpan> pop() {
+            validateDepth();
+            if (depth > 0) {
+                depth--;
+            }
+            return Optional.empty();
+        }
+
+        @Override
+        boolean isEmpty() {
+            validateDepth();
```

Comment:
In theory this call should be safe to delete right? As `pop()` is the only method that can decrement `depth`?

Reply:
Yep, validateDepth only exists to make sure I didn't write a bug :-)